### PR TITLE
Add a check for receiver to validate minimum required fees

### LINKED ISSUE #105: Specify receiver minfeerate
In an environment where transactions with low fees get stuck, the receiver may want to reject requests containing original PSBTs below a minimum fee rate.

Such a check may also deserve a new [well-known error](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-Receivers_well_known_errors) that receivers can return to be displayed to senders to adjust their fees to create a follow-up request if they're willing to accept new fee rate.

Enterprises in particular cannot afford to have inputs stuck in low fee transactions.

### PR DESCRIPTION
**Pull request by jbesraa**
resolves #105 

    Payjoin receiver can set minimum `fee_rate`
    in order to reject payjoin requests that contain
    a psbt which doesnt meet the minimum `fee_rate`
    threshold.

    We add this functionality to `check_broadcast_suitability`
    function which previously was called `check_can_broadcast`.

### CODE REVIEW

File: payjoin/src/receive/mod.rs [hunk 1]
```diff
@@ -347,6 +348,19 @@ impl UncheckedProposal {
         self.psbt.clone().extract_tx()
     }
 
+    /// Check that the Original PSBT meet the minimum feerate receiver requires.
+    pub fn check_if_meet_min_feerate(self, receiver_min_feerate: FeeRate) -> Result<(), Error> {
+        let sender_fee_rate = self.params.min_feerate;
+        // let sender_addtional_fee_contribution = self.params.additional_fee_contribution;
```
Comment (jbesraa) [inline payjoin/src/receive/mod.rs:354]: what else should we take into account in this check?
we have another couple of params that can be interesting:
1) `self.psbt.fee()` (this thing returns `Bitcoin::Amount` but I guess we can convert it to `bitcoin::FeeRate` 
2) `self.params.addition_fee_contr` - this need to be validated(other index input should be set) first which we do in a different check

File: payjoin/src/receive/mod.rs [hunk 2]
```diff
@@ -850,16 +866,17 @@ mod test {
...
         let body = original_psbt.as_bytes();
         let headers = MockHeaders::new(body.len() as u64);
-        UncheckedProposal::from_request(
-            body,
-            "?maxadditionalfeecontribution=182?additionalfeeoutputindex=0",
-            headers,
-        )
+        let min_feerate_param = format!("minfeerate={}", min_feerate.unwrap_or(0));
+        let query = format!(
+            "maxadditionalfeecontribution=182&additionalfeeoutputindex=0&{}",
```
Comment (jbesraa) [inline payjoin/src/receive/mod.rs:871]: this previous query used `?` to separate between pairs, which is not correct. is should be `&` (refer to `url_fromencoded` crate where they split by `&`)

File: payjoin/src/receive/mod.rs (hunk 1 above)
Comment (DanGould) [reply payjoin/src/receive/mod.rs:354]: `params.min_feerate` != original psbt feerate. That parameter can be set arbitrarily. What we want to check is the effective FeeRate of `self.psbt` to make sure this is a reliable fallback if the sender disappears. Otherwise, receiver would be subject to probing attacks where the original psbt fee rate is too low for the receiver to actually ever get it included in a block in the fallback case. As you mention, `additional_fee_contribution` is handled later on and can be ignored in this check.

Perhaps instead of `check_if_meet_min_feerate` this functionality could be combined with `check_can_broadcast`. Since that is already a `testmempoolaccept` check to prevent probing, which is the basis for this issue.

Something like `check_broadcast_suitability(self, min_feerate: Option<bitcoin::FeeRate>, can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>` could replace check_can_broadcast.

min_feerate could be enforced on `assume_interactive_receiver(self, min_feerate: Option<bitcoin::FeeRate>)`` using a shared helper function, too

As a bonus this uses our typestate machine so the check can't be so easily overlooked.

File: payjoin/src/receive/mod.rs (hunk 2 above)
Comment (DanGould) [reply payjoin/src/receive/mod.rs:871]: Nice catch. This should be its own commit

File: payjoin/src/receive/mod.rs (hunk 1 above)
Comment (jbesraa) [reply payjoin/src/receive/mod.rs:354]: sorry for the delay here, working on it

Comment (jbesraa) [reply payjoin/src/receive/mod.rs]: thanks for the patience
I pushed a new version and split the commits.

File: payjoin/src/receive/mod.rs [hunk 3]
```diff
@@ -295,8 +295,9 @@ pub trait Headers {
 ///
 /// If you are implementing an interactive payment processor, you should get extract the original
 /// transaction with extract_tx_to_schedule_broadcast() and schedule, followed by checking
-/// that the transaction can be broadcast with check_can_broadcast. Otherwise it is safe to
+/// that the transaction can be broadcast with check_broadcast_suitability. Otherwise it is safe to
 /// call assume_interactive_receive to proceed with validation.
+#[derive(Debug, Clone)]
```
Comment (jbesraa) [inline payjoin/src/receive/mod.rs:300]: Added for testing/debugging purposes

File: payjoin/src/receive/mod.rs [hunk 4]
```diff
@@ -359,10 +360,18 @@ impl UncheckedProposal {
     /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
     ///
     /// Call this after checking downstream.
-    pub fn check_can_broadcast(
+    pub fn check_broadcast_suitability(
```
Comment (jbesraa) [inline payjoin/src/receive/mod.rs:363]: we probably want to update the documentation of the function to reflect the current change

File: payjoin/src/receive/optional_parameters.rs [hunk 5]
```diff
@@ -4,7 +4,7 @@ use std::fmt;
 use bitcoin::FeeRate;
 use log::warn;
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (jbesraa) [inline payjoin/src/receive/optional_parameters.rs:7]: Added for testing/debugging purposes

File: payjoin/src/receive/error.rs [hunk 6]
```diff
@@ -65,6 +65,10 @@ pub(crate) enum InternalRequestError {
     /// Original PSBT input has been seen before. Only automatic receivers, aka "interactive" in the spec
     /// look out for these to prevent probing attacks.
     InputSeen(bitcoin::OutPoint),
+    /// Fee rate provided by sender is too low
+    /// First argument is the fee rate provided by the sender
+    /// Second argument is the minimum fee rate required by the receiver
+    FeeTooLow(bitcoin::Amount, bitcoin::FeeRate),
```
[the remark this answers is not in the document]
Comment (DanGould) [reply payjoin/src/receive/error.rs:71]: Not sure I understand. Absolute fee and feerate are distinct, no?

[the remark this answers is not in the document]
Comment (jbesraa) [reply payjoin/src/receive/error.rs:71]: yup. you are right. took a second to click

File: payjoin/src/receive/mod.rs [hunk 7]
```diff
@@ -841,7 +841,9 @@ mod test {
         }
     }
 
-    fn proposal_from_test_vector() -> Result<UncheckedProposal, RequestError> {
+    fn proposal_from_test_vector(
+        min_feerate: Option<u64>,
+    ) -> Result<UncheckedProposal, RequestError> {
```
[the remark this answers is not in the document]
Comment (jbesraa) [reply payjoin/src/receive/mod.rs:867]: its used to construct the query a few lines below

File: payjoin/src/receive/mod.rs [hunk 8]
```diff
@@ -850,16 +852,17 @@ mod test {
 
         let body = original_psbt.as_bytes();
         let headers = MockHeaders::new(body.len() as u64);
-        UncheckedProposal::from_request(
-            body,
-            "?maxadditionalfeecontribution=182?additionalfeeoutputindex=0",
```
[the remark this answers is not in the document]
Comment (jbesraa) [reply payjoin/src/receive/mod.rs:855]: its not really needed here.. also not expected by `from_urlencoded`

File: payjoin/src/receive/mod.rs [hunk 9]
```diff
@@ -852,7 +873,7 @@ mod test {
         let headers = MockHeaders::new(body.len() as u64);
         UncheckedProposal::from_request(
             body,
-            "?maxadditionalfeecontribution=182?additionalfeeoutputindex=0",
+            "maxadditionalfeecontribution=182&additionalfeeoutputindex=0&minfeerate=0",
```
Comment (jbesraa) [inline payjoin/src/receive/mod.rs:876]: not required

**Comment by jbesraa** [comment]
@DanGould 
Rebased from master branch which required some minor changes to `v2` stuff, and removed not required changes in the scope of this pull request.
We talked about test vectors for different PSBTs, but I was not able to find a quick solution to generate PSBTs with some configuration regrading fee or weights, might be something we want to develop internally.
I think this pr can go on and we can look into more robust testing in the (near) future.

File: payjoin/src/receive/mod.rs [hunk 10]
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
     }
 
     /// The Sender's Original PSBT
-    pub fn extract_tx_to_schedule_broadcast(&self) -> bitcoin::Transaction {
-        self.psbt.clone().extract_tx()
+    pub fn psbt_transaction(&self) -> bitcoin::Transaction { self.psbt.clone().extract_tx() }
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:349]: This name change seems unrelated to the receiver check this PR is about and I'd prefer it wasn't made. extract_tx (for the specific api reason) tries to convey why the api is available, and extract lets you know it's calculating rather than returning something from memory.

Updating this would also merit a change to the documentation at the front matter of this file

File: payjoin/src/receive/mod.rs [hunk 11]
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
...
+
+    /// The Sender's Original PSBT fee rate.
+    pub fn psbt_fee_rate(&self) -> Result<FeeRate, Error> {
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:352]: Not sure this function should be public, especially since it returns BadRequest, which doesn't make much sense as a return type for a fee_rate here

File: payjoin/src/receive/mod.rs [hunk 12]
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
...
+    pub fn check_broadcast_suitability(
         self,
+        min_fee_rate: Option<FeeRate>,
         can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>,
     ) -> Result<MaybeInputsOwned, Error> {
+        let original_psbt_fee_rate = self.psbt_fee_rate()?;
+        if let Some(min_fee_rate) = min_fee_rate {
+            if original_psbt_fee_rate < min_fee_rate {
+                return Err(Error::BadRequest(
+                    InternalRequestError::PsbtBelowFeeRate(original_psbt_fee_rate, min_fee_rate)
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:384]: I suggest implicit conversion to Error by using `.into()` on its own. See `return Err(InternalRequestError::InvalidContentType(content_type.to_owned()).into());`

File: payjoin/src/receive/mod.rs [hunk 13]
```diff
@@ -873,6 +894,7 @@ mod test {
         use bitcoin::{Address, Network};
 
         let proposal = proposal_from_test_vector().unwrap();
+        assert_eq!(proposal.psbt_fee_rate().unwrap().to_sat_per_vb_floor(), 2);
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:897]: I understand why you might make psbt_fee_rate public for this function, but it's so brief and nonessential to the payjoin api I'd rather see it repeated in the test than exposed publically.

File: payjoin/src/receive/v2.rs [hunk 14]
```diff
@@ -323,10 +323,10 @@ impl Enrolled {
 ///
 /// If you are implementing an interactive payment processor, you should get extract the original
 /// transaction with extract_tx_to_schedule_broadcast() and schedule, followed by checking
-/// that the transaction can be broadcast with check_can_broadcast. Otherwise it is safe to
+/// that the transaction can be broadcast with check_broadcast_suitability. Otherwise it is safe to
 /// call assume_interactive_receive to proceed with validation.
 pub struct UncheckedProposal {
-    inner: super::UncheckedProposal,
+    pub inner: super::UncheckedProposal,
```
Comment (DanGould) [inline payjoin/src/receive/v2.rs:329]: Why make this public? Ideally these typestate machines are kept opaque so that the only modifications happen through the machine.

File: payjoin/src/receive/v2.rs [hunk 15]
```diff
@@ -364,11 +364,11 @@ impl UncheckedProposal {
     /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
     ///
     /// Call this after checking downstream.
-    pub fn check_can_broadcast(
+    pub fn check_broadcast_suitability(
         self,
         can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>,
     ) -> Result<MaybeInputsOwned, Error> {
-        let inner = self.inner.check_can_broadcast(can_broadcast)?;
+        let inner = self.inner.check_broadcast_suitability(None, can_broadcast)?;
```
Comment (DanGould) [inline payjoin/src/receive/v2.rs:371]: This v2 UncheckedProposal should have the same interface for `self.inner.check_boradcast_suitablity` and accept `min_feerate`. Hardcoding`None` should raise a yellow flag.

File: payjoin/src/receive/error.rs [hunk 16]
```diff
@@ -150,6 +157,17 @@ impl fmt::Display for RequestError {
...
+                original_psbt_fee_rate,
+                receiver_min_fee_rate,
+            ) => write_error(
+                f,
+                "original-psbt-rejected",
+                &format!(
+                    "Original PSBT fee rate too low: {} < {}.",
+                    original_psbt_fee_rate, receiver_min_fee_rate
+                ),
+            ),
```
Comment (DanGould) [inline payjoin/src/receive/error.rs:170]: Very nice! 🤌

File: payjoin/src/receive/error.rs [hunk 17]
```diff
@@ -86,6 +86,13 @@ pub(crate) enum InternalRequestError {
     ParsePsbt(bitcoin::psbt::PsbtParseError),
     #[cfg(feature = "v2")]
     Utf8(std::string::FromUtf8Error),
+    /// Original PSBT fee rate is below minimum fee rate set by the receiver.
+    ///
+    /// First argument is the calculated fee rate of the original PSBT.
+    /// [`UncheckedProposal::psbt_fee_rate`]
```
Comment (DanGould) [inline payjoin/src/receive/error.rs:92]: remember to remove this comment if choosing to remove psbt_fee_rate from the public interface.

**Review by DanGould — CHANGES_REQUESTED** [review]
Getting much closer 🔥 I think this gets the job done and only request a few cleanups

File: payjoin/src/receive/mod.rs (hunk 13 above)
Comment (jbesraa) [reply payjoin/src/receive/mod.rs:896]: the pub was a leftover from when i was testing in integration, removed that. 
here in the unit test it can access all the functions/structs in the same module so its fine.

**Comment by jbesraa** [comment]
Thanks for the input @DanGould 
I have addressed all the above

**Comment by DanGould** [comment]
Ci failed to catch the need to change here: https://github.com/payjoin/rust-payjoin/blob/624ab7f0f339d0232644e507b9b66925f7ecfd32/payjoin-cli/src/app.rs#L542

We need to make sure payjoin-cli builds in v2
